### PR TITLE
Add the recreateOption to the object template

### PR DESCRIPTION
**Pull request by mprahl**
When a user needs to update an object's immutable fields, the object must be replaced. The user may opt-in to setting recreateOption on an object template to "IfRequired" and "Always". When set to "IfRequired", normal updates proceed when possible.

Relates:
https://issues.redhat.com/browse/ACM-11846

### CODE REVIEW

File: api/v1/configurationpolicy_types.go [hunk 1]
```diff
@@ -340,6 +342,19 @@ func (c ComplianceType) IsMustNotHave() bool {
 // +kubebuilder:validation:Enum=MustHave;Musthave;musthave;MustOnlyHave;Mustonlyhave;mustonlyhave
 type MetadataComplianceType string
 
+// RecreateOption describes the condition when to delete and recreate an object when an update is required. IfRequired
```
Comment (mprahl) [inline api/v1/configurationpolicy_types.go:345]: @dockerymick could you please review this text?

Comment (yiraeChristineKim) [reply api/v1/configurationpolicy_types.go:345]: Could you add default behaviour?

File: controllers/configurationpolicy_controller.go [hunk 2]
```diff
@@ -2668,32 +2683,70 @@ func (r *ConfigurationPolicyReconciler) checkAndUpdateResource(
...
+						)
+					}
 
-			diff = handleDiff(log, recordDiff, isInform, existingObjectCopy, dryRunUpdatedObj)
+					return true, message, diff, false, nil
+				}
+			} else {
+				removeFieldsForComparison(dryRunUpdatedObj)
+
+				if reflect.DeepEqual(dryRunUpdatedObj.Object, existingObjectCopy.Object) {
```
Comment (yiraeChristineKim) [inline controllers/configurationpolicy_controller.go:2737]: Why is this log mismatch message?

Comment (mprahl) [reply controllers/configurationpolicy_controller.go:2737]: What do you mean?

File: controllers/configurationpolicy_controller.go [hunk 3]
```diff
@@ -2706,18 +2759,56 @@ func (r *ConfigurationPolicyReconciler) checkAndUpdateResource(
...
+
+			for {
+				updatedObj, err = res.Create(context.TODO(), &obj.desiredObj, metav1.CreateOptions{})
+				if !k8serrors.IsAlreadyExists(err) {
+					// If there is no error or the error is unexpected, break for the error handling below
+					break
+				}
+
+				if time.Since(start) > time.Second*10 {
+					message = fmt.Sprintf(
```
Comment (yiraeChristineKim) [inline controllers/configurationpolicy_controller.go:2790]: In this for loop, it doesn't delete the obj. so the message should be changed?

**Review by yiraeChristineKim — APPROVED** [review]

File: controllers/configurationpolicy_controller.go [hunk 4]
```diff
@@ -1791,15 +1806,18 @@ func (r *ConfigurationPolicyReconciler) handleSingleObj(
 			throwSpecViolation, msg, diff, triedUpdate, updatedObj = r.checkAndUpdateResource(
 				obj, objectT, remediation,
 			)
+
+			if updatedObj != nil && string(updatedObj.GetUID()) != uid {
+				uid = string(updatedObj.GetUID())
+				created = true
+			}
```
[the remark this answers is not in the document]
Comment (mprahl) [reply controllers/configurationpolicy_controller.go:1813]: Yes, exactly. So if the UID changes, the objectProperties will get the new UID.

File: api/v1/configurationpolicy_types.go [hunk 5]
```diff
@@ -340,6 +342,19 @@ func (c ComplianceType) IsMustNotHave() bool {
...
+// will recreate the object when updating an immutable field. Always will always recreate the object if a mismatch is
+// detected. RecreateOption has no effect when the remediationAction is inform. IfRequired has no effect on clusters
+// without dry run update support. Default is None.
+// +kubebuilder:validation:Enum=None;IfRequired;Always
+type RecreateOption string
```
[the remark this answers is not in the document]
Comment (mprahl) [reply api/v1/configurationpolicy_types.go]: This is the code that needs updating but it's not a big deal. I have the local changes queued up. Just need to run the tests once this merges:
https://github.com/open-cluster-management-io/governance-policy-framework-addon/blob/f299b7020823b3a09ab872ea4731080b42ff09d8/controllers/templatesync/template_sync.go#L959-L967

[the remark this answers is not in the document]
Comment (JustinKuli) [reply api/v1/configurationpolicy_types.go]: Reading that makes me think that template-sync might be thinking it needs to do more updates to OperatorPolicy than it does. I thought I had checked and it wasn't stuck in a loop at least, but I'll need to investigate more

**Pull request merged**